composer/utils/checkpoint.py

```diff
@@ -217,17 +217,30 @@ def read_data(self, plan: LoadPlan, planner: LoadPlanner):
         first_replica = self.device_mesh is None or self.device_mesh.ndim == 1 or (
             self.device_mesh.ndim >= 2 and self.device_mesh.get_local_rank(mesh_dim=0) == 0)
 
-        # 1. Download to the destination all files this rank needs if on first replica
+        # 1. Collect the relative paths to download for all ranks for deduplication
+        relative_file_paths = set()
+        for plan_item in plan.items:
+            relative_file_paths.add(self.storage_data[plan_item.storage_index].relative_path)
+        all_file_paths = dist.all_gather_object(relative_file_paths)
+        log.debug(f"Files required for all ranks {all_file_paths}")
+
+        # 2. Download to the destination all files this rank needs if on first replica
         if first_replica:
             log.debug(f'Rank {dist.get_global_rank()} starting to download files.')
+
+            # Download the files for this rank if they haven't been downloaded before
+            rank = dist.get_local_rank()
             for plan_item in plan.items:
-                # Each plan item has a storage index which points to the relative path of the shard file at save time.
                 relative_file_path = self.storage_data[plan_item.storage_index].relative_path
+                # Check if the file has already been downloaded by another rank
+                is_downloaded = any(relative_file_path in all_file_paths[i] for i in range(rank))
```

Comment:
Not sure I understand how this works. `all_file_paths[i]` is a list of all the files a rank needs not a list of files the rank already downloaded?

Reply:
comment/variable name can be more clear. 

Idea is another rank is responsible for downloading it. Eg if rank 9 and 10 both need shard 0, only the lower numbered rank (9) downloads it instead of both

Reply:
oh I see. it only loops up to the ranks before the current rank. that's almost too slick

Reply:
Oh yep I'll update the comment to clarify